File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.io.*;
import java.util.*;

public class Main {
    public static int[] arr;
    public static Integer[] dp;
    public static void main(String[] args) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(reader.readLine());
        arr = Arrays.stream(reader.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        dp = new Integer[n];
        for (int i = 0; i < arr.length; i++) {
            getLength(i);
        }
        int max = Integer.MIN_VALUE;
        for (int i = 0; i < dp.length; i++) max = Math.max(max, dp[i]);
        System.out.println(max);
        reader.close();
    }
    private static int getLength(int n) {
        if (dp[n] == null) {
            dp[n] = 1;
            for (int i = n - 1; i > -1; i--) {
                if(arr[i] < arr[n]) dp[n] = Math.max(dp[n], getLength(i) + 1);
            }
        }
        return dp[n];
    }
}

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.py
import sys
input = sys.stdin.readline

n = int(input().rstrip())
arr = list(map(int, input().rstrip().split()))
result = [1 for _ in range(n)]

def fill(n):
	global result
	if result[n] == 1:
		for i in range(n - 1, -1, -1):
			if arr[i] < arr[n]: result[n] = max(result[n], fill(i) + 1)
	return result[n]

for i in range(n): fill(i)
print(max(result))
